File: app/src/main/java/com/jcr/sharedtasks/ui/MainActivityViewModel.java
package com.jcr.sharedtasks.ui;

import android.arch.lifecycle.LiveData;
import android.arch.lifecycle.MutableLiveData;
import android.arch.lifecycle.Transformations;
import android.arch.lifecycle.ViewModel;
import android.net.Uri;
import android.support.annotation.VisibleForTesting;

import com.google.firebase.auth.FirebaseUser;
import com.jcr.sharedtasks.model.Project;
import com.jcr.sharedtasks.model.ProjectReference;
import com.jcr.sharedtasks.model.Task;
import com.jcr.sharedtasks.repository.ProjectsRepository;
import com.jcr.sharedtasks.repository.SignInRepository;
import com.jcr.sharedtasks.util.AbsentLiveData;
import com.jcr.sharedtasks.util.DeepLinkUtils;

import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

import javax.inject.Inject;

public class MainActivityViewModel extends ViewModel {

    @VisibleForTesting
    final MutableLiveData<Boolean> logged;
    private final LiveData<List<ProjectReference>> projectsReferences;
    private final SignInRepository signInRepository;
    private final ProjectsRepository projectsRepository;

    @Inject
    public MainActivityViewModel(SignInRepository signInRepository, ProjectsRepository projectsRepository) {
        this.logged = new MutableLiveData<>();
        this.signInRepository = signInRepository;
        this.projectsRepository = projectsRepository;
        projectsReferences = Transformations.switchMap(logged, logged -> {
            if (logged) {
                return projectsRepository.getProjectsReferences();
            } else {
                return AbsentLiveData.create();
            }
        });
    }

    public void onSignedInitialize(FirebaseUser user) {
        this.signInRepository.onSignedInitialize(user);
        logged.setValue(true);
    }

    public boolean isUserLogged() {
        return logged.getValue() != null && logged.getValue();
    }

    public LiveData<List<ProjectReference>> getProjectUUIDs() {
        return projectsReferences;
    }

    public String parseDeeplink(Uri data) {
        ProjectReference invitedProjectReference = DeepLinkUtils.parseProjectUUID(data);
        projectsRepository.createProjectReference(invitedProjectReference);
        return invitedProjectReference.getProjectUUID();
    }
}

File: app/src/main/java/com/jcr/sharedtasks/ui/MainActivityViewModel.kt
package com.jcr.sharedtasks.ui

import android.net.Uri
import androidx.annotation.VisibleForTesting
import androidx.lifecycle.*

import com.google.firebase.auth.FirebaseUser
import com.jcr.sharedtasks.model.Project
import com.jcr.sharedtasks.model.ProjectReference
import com.jcr.sharedtasks.model.Task
import com.jcr.sharedtasks.repository.ProjectsRepository
import com.jcr.sharedtasks.repository.SignInRepository
import com.jcr.sharedtasks.util.AbsentLiveData
import com.jcr.sharedtasks.util.DeepLinkUtils
import kotlinx.coroutines.flow.*

import java.util.ArrayList
import java.util.UUID

import javax.inject.Inject

class MainActivityViewModel @Inject
constructor(private val signInRepository: SignInRepository, private val projectsRepository: ProjectsRepository) : ViewModel() {

    @VisibleForTesting
    private val logged = MutableStateFlow<Boolean?>(null)
    val projectUUIDs: StateFlow<List<ProjectReference>> = logged.transform { logged ->
        if (logged == true) {
            emitAll(projectsRepository.projectsReferences)
        }
    }.stateIn(
        scope = viewModelScope,
        started = SharingStarted.WhileSubscribed(5_000),
        initialValue = emptyList()
    )

    val isUserLogged: Boolean
        get() = logged.value != null && logged.value!!

    fun onSignedInitialize(user: FirebaseUser) {
        signInRepository.onSignedInitialize(user)
        logged.value = true
    }

    fun parseDeeplink(data: Uri): String {
        val invitedProjectReference = DeepLinkUtils.parseProjectUUID(data)
        projectsRepository.createProjectReference(invitedProjectReference)
        return invitedProjectReference.projectUUID
    }
}
